src/mlpack/core/tree/octree/octree_impl.hpp

```diff
@@ -0,0 +1,854 @@
+void Octree<MetricType, StatisticType, MatType>::Serialize(
+    Archive& ar,
+    const unsigned int /* version */)
+{
+  using data::CreateNVP;
+
+  // 
+
+  // If we're loading and we have children, they need to be deleted.
+  if (Archive::is_loading::value)
+  {
+    for (size_t i = 0; i < children.size(); ++i)
+      delete children[i];
+    children.clear();
+
+    if (!parent)
+      delete dataset;
+  }
+
+  ar & CreateNVP(begin, "begin");
+  ar & CreateNVP(count, "count");
+  ar & CreateNVP(bound, "bound");
+  ar & CreateNVP(stat, "stat");
+  ar & CreateNVP(parentDistance, "parentDistance");
+  ar & CreateNVP(furthestDescendantDistance, "furthestDescendantDistance");
+  ar & CreateNVP(metric, "metric");
+
+  // Due to quirks of boost::serialization, depending on how the user
+  // serializes the tree, it's possible that the root of the tree will
+  // accidentally be serialized twice.  So if we are a first-level child, we
+  // avoid serializing the parent.  The true (non-duplicated) parent will fix
+  // the parent link.
+  bool hasFakeParent = false;
+  if (Archive::is_saving::value && parent != NULL && parent->parent == NULL)
+  {
+    Octree* fakeParent = NULL;
+    hasFakeParent = true;
+    ar & CreateNVP(fakeParent, "parent");
+    ar & CreateNVP(hasFakeParent, "hasFakeParent");
+  }
+  else
+  {
+    ar & CreateNVP(parent, "parent");
+    ar & CreateNVP(hasFakeParent, "hasFakeParent");
+  }
+
+  // Only serialize the dataset if we don't have a fake parent.  Otherwise, the
+  // real parent will come and set it later.
+  if (!hasFakeParent)
+    ar & CreateNVP(dataset, "dataset");
+
+  size_t numChildren = 0;
+  if (Archive::is_saving::value)
+    numChildren = children.size();
+  ar & CreateNVP(numChildren, "numChildren");
+  if (Archive::is_loading::value)
+    children.resize(numChildren);
+
+  for (size_t i = 0; i < numChildren; ++i)
+  {
+    std::ostringstream oss;
+    oss << "child" << i;
+    ar & CreateNVP(children[i], oss.str());
+  }
+
+  // Fix the child pointers, if they were set to a fake parent.
+  if (Archive::is_loading::value && parent == NULL)
+  {
+    for (size_t i = 0; i < children.size(); ++i)
+    {
+      children[i]->dataset = this->dataset;
+      children[i]->parent = this;
+    }
+  }
+}
+
+//! Split the node.
+template<typename MetricType, typename StatisticType, typename MatType>
+void Octree<MetricType, StatisticType, MatType>::SplitNode(
+    const arma::vec& center,
+    const double width,
+    const size_t maxLeafSize)
+{
+  // No need to split if we have fewer than the maximum number of points in this
+  // node.
+  if (count <= maxLeafSize)
+    return;
+
+  // We must split the dataset by sequentially creating each of the children.
+  // We do this in two steps: first we make a pass to count the number of points
+  // that will fall into each child; then in the second pass we rearrange the
+  // points and create the children.
+  arma::Col<size_t> childCounts(std::pow(2, dataset->n_rows),
```

Comment:
I think that `((size_t) 2) << dataset->n_rows` is better than `std::pow()`, how do you think?